### ISSUE TITLE
`asyncio.BaseProactorEventLoop` feeds bytearray objects to the `Protocol.data_received` callback, contradicting docs

### ISSUE THREAD

**Issue by jonathanslenders** [python/cpython#123496]
# Bug report

### Bug description:

The documentation says: https://docs.python.org/3/library/asyncio-protocol.html#asyncio.Protocol.data_received :

```
Called when some data is received. data is a non-empty bytes object containing the incoming data.
```

So, it says it should be a bytes object. However:

- First, `self._data` is a `bytearray` here: https://github.com/python/cpython/blob/f8a736b8e14ab839e1193cb1d3955b61c316d048/Lib/asyncio/proactor_events.py#L191
- This gets passed to `self._data_received here: https://github.com/python/cpython/blob/f8a736b8e14ab839e1193cb1d3955b61c316d048/Lib/asyncio/proactor_events.py#L230
- The passed into `Protocol.data_received` here: https://github.com/python/cpython/blob/f8a736b8e14ab839e1193cb1d3955b61c316d048/Lib/asyncio/proactor_events.py#L274

So, there we get a bytearray object.

This conflicts with the current type hints from typeshed: https://github.com/python/typeshed/blob/main/stdlib/asyncio/protocols.pyi#L14 as well as with the docs.

Is this intentional or not?

The patch is simple:

```diff
--- a/Lib/asyncio/proact
+++ b/Lib/asyncio/proactor_events.py
@@ -271,7 +271,7 @@ def _data_received(self, data, length):
                                   'call failed.')
                 return
         else:
-            self._protocol.data_received(data)
+            self._protocol.data_received(bytes(data))

     def _loop_reading(s
         length = -1
```

Related typeshed issue: https://github.com/python/typeshed/issues/12606

### CPython versions tested on:

3.13

### Operating systems tested on:

Windows

**Comment by jonathanslenders** [comment]
Actually, maybe it's a duplicate of: https://github.com/python/cpython/issues/99941

**Comment by jonathanslenders** [comment]
Closing this. Apparently it was fixed in v3.12.0a thanks to https://github.com/python/cpython/commit/1bb68ba6d9de6bb7f00aee11d135123163f15887

**Comment by agronholm** [comment]
I can still reproduce this with 3.12.5.
